README.md

Facebook, Twitter, Yahoo, and Bing are constantly trying to view your website... but they don't execute javascript. That's why Prerender was built. Prerendering is useful for AngularJS SEO, BackboneJS SEO, EmberJS SEO, and any other javascript framework.

Behind the scenes, Prerender is a node server forked from [prerender.io](http://prerender.io) that uses phantomjs to create static HTML out of a javascript page. Netlify hosts this as a service which you can configure in our UI to work automatically with your page.

Prerender adheres to google's `_escaped_fragment_` proposal, which we recommend you use. It's easy:
- Just add &lt;meta name="fragment" content="!"> to the &lt;head> of all of your pages
- If you use hash urls (#), change them to the hash-bang (#!)
- Turn on prerendering in the Netlify Build & Deploy Settings
- That's it! Perfect SEO on javascript pages.

This package starts multiple phantomjs processes to maximize throughput, this may be a bit overkill for your testing so you might not want to leave it running after testing is complete.


## How it works
This is a simple service that only takes a url and returns the rendered HTML (with all script tags removed).

Note: you should proxy the request through your server (using middleware) so that any relative links to CSS/images/etc still work.  If you don't, you'll see the relative links fail but the HTML output will still be useful for debugging purposes.

## Running locally
If you are trying to test Prerender with your website on localhost, you'll have to run the Prerender server locally so that Prerender can access your local dev website.

	$ git clone https://github.com/netlify/prerender.git
	$ cd prerender
	$ npm install
	$ node server.js

Prerender will now be running on http://localhost:3000 and you can test your page (deployed on netlify or elsewhere!) using links like this:

http://localhost:3000/https://www.netlify.com/pricing

You will see debug output in the shell you're running the service in that will hopefully enable you to transpile your javascript appropriately to comply with PhantomJS' limitations.


## Options

### workers
```
var prerender = require('./lib');

var server = prerender({
    workers: 1
});

server.start();
```

The number of Prerender workers that you'd like to start. We suggest 1 per CPU on your machine. `Default: os.cpus().length`

### iterations
```
var prerender = require('./lib');

var server = prerender({
    iterations: 40
});

server.start();
```

The number of pages Prerender should render before restarting the worker. WARNING: Will restart Prerender mid-request if two requests are in flight, causing a 504 response to be sent.

Shutting Prerender down reclaims memory to ensure good performance. You can also set the environment variable of `NUM_ITERATIONS` instead of passing in the `iterations` parameter. `Default: 40`

### softIterations
```
var prerender = require('./lib');

var server = prerender({
    softIterations: 5
});

server.start();
```

The number of pages Prerender should render before restarting the worker. This option counts the number of requests in flight and only restarts the worker if no requests are in flight. If you constantly have more than 1 request in flight, this won't restart the server.

Shutting Prerender down reclaims memory to ensure good performance. You can also set the environment variable of `NUM_SOFT_ITERATIONS` instead of passing in the `softIterations` parameter. `Default: 30`

### cookiesEnabled
```
var prerender = require('./lib');

var server = prerender({
    cookiesEnabled: true
});

server.start();
```

If Prerender should use Cookies. You can also set the environment variable of `COOKIES_ENABLED` instead of passing in the `cookiesEnabled` parameter. `Default: false`

### logRequests
```
var prerender = require('./lib');

var server = prerender({
    logRequests: true
});

server.start();
```

Causes the Prerender server to print out every request made represented by a `+` and every response received represented by a `-`. Lets you analyze page load times. `Default: false`

### pageDoneCheckTimeout
```
var prerender = require('./lib');

var server = prerender({
    pageDoneCheckTimeout: 300
});

server.start();
```

Number of milliseconds between the interval of checking whether the page is done loading or not. You can also set the environment variable of `PAGE_DONE_CHECK_TIMEOUT` instead of passing in the `pageDoneCheckTimeout` parameter. `Default: 300`

### resourceDownloadTimeout
```
var prerender = require('./lib');

var server = prerender({
    resourceDownloadTimeout: 10000
});

server.start();
```

Number of milliseconds to wait while downloading the page, waiting for all pending requests/ajax calls to complete before timing out and continuing on. Time out condition does not cause an error, it just moves on to the javascript execution stage. You can also set the environment variable of `RESOURCE_DOWNLOAD_TIMEOUT` instead of passing in the `resourceDownloadTimeout` parameter. `Default: 10000`

### waitAfterLastRequest
```
var prerender = require('./lib');

var server = prerender({
    waitAfterLastRequest: 500
});

server.start();
```

Number of milliseconds to wait after the number of requests/ajax calls in flight reaches zero. Javascript execution begins after this in order to pull the HTML off of the page. You can also set the environment variable of `WAIT_AFTER_LAST_REQUEST` instead of passing in the `waitAfterLastRequest` parameter. `Default: 500`

### jsTimeout
```
var prerender = require('./lib');

var server = prerender({
    jsTimeout: 10000
});

server.start();
```

Number of milliseconds to continue trying to pull the HTML off of the page using javascript before timing out. Once the timeout is hit, Prerender returns a 200 response with the last HTML that it was able to pull off of the page. You can also set the environment variable of `JS_TIMEOUT` instead of passing in the `jsTimeout` parameter. `Default: 10000`

### jsCheckTimeout
```
var prerender = require('./lib');

var server = prerender({
    jsCheckTimeout: 300
});

server.start();
```

Number of milliseconds between the interval of checking whether the javascript timeout has been reached or not. You can also set the environment variable of `JS_CHECK_TIMEOUT` instead of passing in the `jsCheckTimeout` parameter. `Default: 300`

### noJsExecutionTimeout
```
var prerender = require('./lib');

var server = prerender({
    noJsExecutionTimeout: 3000
});

server.start();
```

Number of milliseconds to wait while not being able to execute javascript before determining that the Prerender server hasn't been able to execute javascript, usually due to a webpage not giving up control of the JS execution thread (infinite loop). You can also set the environment variable of `NO_JS_EXECUTION_TIMEOUT` instead of passing in the `noJsExecutionTimeout` parameter. `Default: 3000`

### evaluateJavascriptCheckTimeout
```
var prerender = require('./lib');

var server = prerender({
    evaluateJavascriptCheckTimeout: 300
});

server.start();
```

Number of milliseconds between executing the javascript on the webpage to pull off the HTML. Pulling off the HTML only happens multiple times when `window.prerenderReady` is set to false. You can also set the environment variable of `EVALUATE_JAVASCRIPT_CHECK_TIMEOUT` instead of passing in the `evaluateJavascriptCheckTimeout` parameter. `Default: 300`

## Plugins

We use a plugin system in the same way that Connect and Express use middleware. Our plugins are a little different and we don't want to confuse the prerender plugins with the [prerender middleware](#middleware), so we opted to call them "plugins".

Plugins are in the `lib/plugins` directory, and add functionality to the prerender service.

Each plugin can implement any of the plugin methods:

####`init()`

####`beforePhantomRequest(req, res, next)`

####`onPhantomPageCreate(phantom, req, res, next)`

####`afterPhantomRequest(req, res, next)`

####`beforeSend(req, res, next)`

## Available plugins

You can enable the plugins in `server.js` by uncommenting the corresponding lines.

### basicAuth

If you want to only allow access to your Prerender server from authorized parties, enable the basic auth plugin.

You will need to add the `BASIC_AUTH_USERNAME` and `BASIC_AUTH_PASSWORD` environment variables.
```
export BASIC_AUTH_USERNAME=prerender
export BASIC_AUTH_PASSWORD=test
```

Then make sure to pass the basic authentication headers (password base64 encoded).

```
curl -u prerender:wrong http://localhost:1337/http://example.com -> 401
curl -u prerender:test http://localhost:1337/http://example.com -> 200
```

### removeScriptTags

We remove script tags because we don't want any framework specific routing/rendering to happen on the rendered HTML once it's executed by the crawler. The crawlers may not execute javascript, but we'd rather be safe than have something get screwed up.

For example, if you rendered the HTML of an angular page but left the angular scripts in there, your browser would try to execute the angular routing and rendering on a page that no longer has any angular bindings.

This plugin implements the `beforeSend` funtion, therefore cached HTML pages still contain scripts tags until they get served.

### httpHeaders

If your Javascript routing has a catch-all for things like 404's, you can tell the prerender service to serve a 404 to google instead of a 200. This way, google won't index your 404's.

Add these tags in the `<head>` of your page if you want to serve soft http headers. Note: Prerender will still send the HTML of the page. This just modifies the status code and headers being sent.

Example: telling prerender to server this page as a 404
```html
<meta name="prerender-status-code" content="404">
```

Example: telling prerender to serve this page as a 302 redirect
```html
<meta name="prerender-status-code" content="302">
<meta name="prerender-header" content="Location: http://www.google.com">
```

### whitelist

If you only want to allow requests to a certain domain, use this plugin to cause a 404 for any other domains.

You can add the whitelisted domains to the plugin itself, or use the `ALLOWED_DOMAINS` environment variable.

`export ALLOWED_DOMAINS=www.prerender.io,prerender.io`

### blacklist

If you want to disallow requests to a certain domain, use this plugin to cause a 404 for the domains.

You can add the blacklisted domains to the plugin itself, or use the `BLACKLISTED_DOMAINS` environment variable.

`export BLACKLISTED_DOMAINS=yahoo.com,www.google.com`


### <a id='s3-html-cache'></a>
### s3HtmlCache

A `GET` request will check S3 for a cached copy. If a cached copy is found, it will return that. Otherwise, it will make the request to your server and then persist the HTML to the S3 cache.

A `POST` request will skip the S3 cache. It will make a request to your server and then persist the HTML to the S3 cache. The `POST` is meant to update the cache.

You'll need to sign up with Amazon Web Services and export these 3 environment variables.

```
$ export AWS_ACCESS_KEY_ID=<aws access key>
$ export AWS_SECRET_ACCESS_KEY=<aws secret access key>
$ export S3_BUCKET_NAME=<bucket name>
```

Warning! Your keys should be kept private and you'll be charged for all files uploaded to S3.

> If Prerender is hosted on a EC2 instance, you can also take advantage of [IAM instance roles](http://aws.typepad.com/aws/2012/06/iam-roles-for-ec2-instances-simplified-secure-access-to-aws-service-apis-from-ec2.html)
so that you don't need to export your AWS credentials.

> You can also export the S3_PREFIX_KEY variable so that the key (which is by default the complete requested URL) is
prefixed. This is useful if you want to organize the snapshots in the same bucket.

#### Region 


By default, s3HtmlCache works with the US Standard region (East), if your bucket is localized in another region you can config it with an environment variable : `AWS_REGION`.

```
$ export AWS_REGION=<region name>
```

For example :

```
$ export AWS_REGION=eu-west-1
```

### inMemoryHtmlCache

*Note* The in memory cache is per process so if you have multiple Prerender workers then they do not share a cache. For higher traffic websites, use a common cache like redis.

An in memory cache but you can easily change it to any caching system compatible with the `cache-manager` nodejs package.

For example, with the request:

`GET http://service.prerender.io/https://www.facebook.com/`

First time: Overall Elapsed:	00:00:03.3174661

With cache: Overall Elapsed:	00:00:00.0360119


### logger

This will show console.log's from the phantomjs page in your local console. Great for debugging.

### mongodbCache

Caches pages in a MongoDB database. Available at [prerender-mongodb-cache](https://github.com/lammertw/prerender-mongodb-cache) by [@lammertw](https://github.com/lammertw)


### memjsCache

Caches pages in a memjs(memcache) service. Available at [prerender-memjs-cache](https://github.com/lammertw/prerender-memjs-cache) by [@lammertw](https://github.com/lammertw)


### levelCache

Caches pages in a levelDB database. Available at [prerender-level-cache](https://github.com/maxlath/prerender-level-cache) by [@maxlath](https://github.com/maxlath)

### accessLog

Create access log file for prerendered requests. Available at [prerender-access-log](https://github.com/unDemian/prerender-access-log) by [@unDemian](https://github.com/unDemian)


## License

The MIT License (MIT)

Copyright (c) 2013 Todd Hooper &lt;todd@prerender.io&gt;

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## 5.0.0 - 2018-02-28
### Changed
- Using headless Chrome version of [prerender](https://github.com/prerender/prerender) now.

## 4.4.1 - 2016-12-28
### Changed
- Whoops. Make sure `shouldEncodeURLBeforeBrowserFetch` defaults to true.

## 4.4.0 - 2016-12-28
### Added
- Added `shouldEncodeURLBeforeBrowserFetch` to allow projects that use prerender to determine whether they want to call `encodeURI` on the URL before fetching it in PhantomJS. Useful for some URLs that might have encoded slashes in them, since encoding them further would cause incorrect behavior.

## 4.3.1 - 2016-08-25
### Changed
- Fixed issue where PhantomJS crashed and then disposing caused bad phantomjs state

## 4.3.0 - 2016-08-04
### Changed
- Bumped all dependency versions to latest

## 4.2.0 - 2016-08-04
### Added
- Added ability for cluster master to kill last known phantomjs pid for a worker if the worker dies (preventing orphaned phantomjs instances)
### Changed
- Better terminating for cluster workers

## 4.1.0 - 2016-07-27
### Added
- Added NUM_SOFT_ITERATIONS to try to kill phantomjs and reclaim memory when no requests are in flight. This should be set to a low number (1-10) so that PhantomJS can be restarted often when it isn't doing anything else. NUM_ITERATIONS should still be set to something like 40-50 to make sure to force kill PhantomJS even if a request is in flight.
- Added clearing of memory cache to prevent PhantomJS from returning a 304
### Changed
- Fixed issue where prerender-status-code set to `200` was causing the page to skip being cached
- Fixed an issue where we weren't using the correct pid when trying to force kill PhantomJS.
- Moved clearing of memory cache and local storage to before the page loads instead of after. This will prevent edge cases that could cause a 304.

## 4.0.10 - 2016-06-01
### Changed
- Fixed issue where S3HtmlCache was calling next() before finishing saving to the cache

## 4.0.9 - 2016-05-08
### Changed
- Fixed issue where we were calling `hasOwnProperty` on a `querystring` that no longer had Object on it's prototype chain

## 4.0.8 - 2016-03-24
### Changed
- Fixed issue where a webpage calling window.close would cause Prerender to be unable to shutdown PhantomJS properly

## 4.0.7 - 2016-03-22
### Changed
- S3 cache plugin was incorrectly saving non-200 status code responses to the cache

## 4.0.6 - 2016-03-09
### Changed
- preserve phantom arguments when server is restarting
- use default when phantomArguments is empty

## 4.0.5 - 2016-02-29
### Changed
- prevent multiple phantomjs instances from being started with low number of iterations
- try to check to see if phantomjs has actually been disposed. if not, force kill it.

## 4.0.4 - 2016-02-18
### Changed
- added engines to package.json and fixed possible bug in checking options passed in
- prevent weird hangup on error setting a header with a newline
- make sure we catch any errors thrown from phridge and continue
- kill workers (and phantomjs) on SIGTERM

## 4.0.3 - 2016-02-12
### Added plugin to send a header of X-Prerender: 1 with every request

## 4.0.2 - 2016-02-12
#### Now using PhantomJS version 2.1
### Changed
- Changed `PHANTOM_CLUSTER_NUM_WORKERS` to `PRERENDER_NUM_WORKERS` in server.js
- Changed `PHANTOM_WORKER_ITERATIONS` to `PRERENDER_NUM_ITERATIONS` in server.js
- Switched from `phantomjs-node` bridge to `phridge`
 - All Prerender plugins that access PhantomJS need to be rewritten to support new [phridge](https://github.com/peerigon/phridge) integration.
 For example, change this:
 ```
 req.prerender.page.set('onConsoleMessage', function(msg) {
     console.log(msg);
 });
 ```
 to this:
 ```
 req.prerender.page.run(function() {

 	this.onConsoleMessage = function(msg) {
            console.log(msg);
        };
 });
 ```
 Please see [phridge](https://github.com/peerigon/phridge) for more info on how to interact with PhantomJS through `phridge`.

 ###Removed
 - Removed `PHANTOM_CLUSTER_BASE_PORT` since `phridge` doesn't start a webserve to talk to PhantomJS, so it's no longer needed.
 - Removed `PHANTOM_CLUSTER_MESSAGE_TIMEOUT` since `phridge` doesn't start a webserve to talk to PhantomJS, so it's no longer needed.


server.js
"use strict";

const http = require("http");
const util = require("./lib/util");
const server = require("prerender/lib/server");
const express = require("express");
const app = express();
const bodyParser = require("body-parser");
const compression = require("compression");
const morgan = require("morgan");

server.init({});
server.onRequest = server.onRequest.bind(server);

morgan.token("request_id", function(req, res) {
  const rhdr = process.env.REQUEST_ID_HEADER;
  if (rhdr && req.headers[rhdr]) {
    return req.headers[rhdr];
  }
  return "unknown";
});

morgan.token("cache_type", function(req, res) {
  if (req.prerender) {
    return req.prerender.cacheHit ? "CACHE_HIT" : "CACHE_MISS";
  }
  return "-";
});

morgan.token("prerender_url", function(req, res) {
  return (req.prerender && req.prerender.url) || "-";
});

app.use(
  morgan(
    `:date[iso] method=:method status=:status url=:url prerender_url=:prerender_url cache_type=:cache_type timing=:response-time[0] referrer=":referrer" user_agent=":user-agent" request_id=:request_id`
  )
);
app.disable("x-powered-by");
app.use(compression());

app.get("*", server.onRequest);

//dont check content-type and just always try to parse body as json
app.post("*", bodyParser.json({ type: () => true }), server.onRequest);

app.use(function(err, req, res, next) {
  res.status(500).end();
  util.log(`Unhandled request error error=${err} stack=${err.stack}`);
});

server.use(require("./lib/plugins/healthCheckAuth"));
server.use(require("prerender/lib/plugins/basicAuth"));
server.use(require("prerender/lib/plugins/blockResources"));
server.use(require("prerender/lib/plugins/blacklist"));
server.use(require("prerender/lib/plugins/removeScriptTags"));
server.use(require("prerender/lib/plugins/httpHeaders"));
server.use(require("./lib/plugins/s3HtmlCache"));

server.start();

const port = process.env.PORT || 3000;
app.listen(port, () =>
  util.log(`Prerender server accepting requests on port ${port}`)
);


lib/plugins/s3HtmlCache.js
const cacheManager = require("cache-manager");
const AWS = require("aws-sdk");
const util = require("../util");

module.exports = {
  init() {
    const s3 = new AWS.S3({ params: { Bucket: process.env.S3_BUCKET_NAME } });
    this.prefix = process.env.S3_PREFIX_KEY;
    this.expirations = [];
    if (process.env.PRERENDER_EXPIRES_DURATION) {
      const values = process.env.PRERENDER_EXPIRES_DURATION.split(";");
      values.forEach(v => {
        this.expirations.push(v.split("="));
      });
    }

    this.cache = cacheManager.caching({
      store: new S3Cache(s3)
    });
  },

  _prefixedKey(key) {
    if (this.prefix) {
      return `${this.prefix}/${key}`;
    }
    return key;
  },

  requestReceived(req, res, next) {
    if (req.method !== "GET") {
      return next();
    }

    this.cache.get(this._prefixedKey(req.prerender.url), function(err, result) {
      if (!err && result) {
        req.prerender.cacheHit = true;
        return res.send(200, result);
      }

      if (err && err.statusCode !== 404 && err.statusCode !== 403) {
        util.log(`S3 GET failed error="${err}" url="${req.prerender.url}"`);
      }
      return next();
    });
  },

  beforeSend(req, res, next) {
    if (req.prerender.statusCode !== 200 || req.prerender.cacheHit) {
      return next();
    }

    const pkey = this._prefixedKey(req.prerender.url);
    const opts = {};
    this.expirations.some(kv => {
      const [expKey, ttl] = kv;
      const matches = pkey.indexOf(expKey) != -1;
      if (matches) {
        opts.ttl = Number(ttl);
        util.log(`host_match=${expKey} url="${pkey}" ttl=${opts.ttl}`);
      }
      return matches;
    });

    this.cache.set(pkey, req.prerender.content, opts, function(err) {
      if (err) {
        util.log(`S3 PUT failed error="${err}" url="${req.prerender.url}"`);
      }
    });

    return next();
  }
};

class S3Cache {
  constructor(s3) {
    this.s3 = s3;
  }

  get(key, opts, cb) {
    if (typeof opts === "function") {
      cb = opts;
      opts = {};
    }
    this.s3
      .getObject({ Key: key })
      .promise()
      .then(data => {
        cb(null, data.Body);
      })
      .catch(err => {
        cb(err);
      });
  }

  set(key, value, opts, cb) {
    if (typeof opts === "function") {
      cb = opts;
      opts = {};
    }

    const params = {
      Key: key,
      ContentType: "text/html;charset=UTF-8",
      StorageClass: "REDUCED_REDUNDANCY",
      Body: value
    };

    if (opts.ttl) {
      const now = Date.now();
      now.setSeconds(now.getSeconds() + Number(opts.ttl));
      params["Expires"] = now;
    }
    this.s3.putObject(params, cb);
  }
}
